### ISSUE TITLE
json module ref leaks detected by test_json

### ISSUE THREAD

**Issue by 6ccb8c66-b6ca-4c37-b449-e35b11a6b9e4** [python/cpython#75119]
BPO | [30936](https://bugs.python.org/issue30936)
--- | :---
Nosy | @vstinner, @serhiy-storchaka, @segevfiner
PRs | <li>python/cpython#2712</li><li>python/cpython#2727</li><li>python/cpython#2728</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-07-16.05:15:26.821>
created_at = <Date 2017-07-15.10:43:00.350>
labels = ['extension-modules', '3.7', 'performance']
title = 'json module ref leaks detected by test_json'
updated_at = <Date 2017-07-16.14:14:47.513>
user = 'https://github.com/segevfiner'
```

bugs.python.org fields:
```python
activity = <Date 2017-07-16.14:14:47.513>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-07-16.05:15:26.821>
closer = 'serhiy.storchaka'
components = ['Extension Modules']
creation = <Date 2017-07-15.10:43:00.350>
creator = 'Segev Finer'
dependencies = []
files = []
hgrepos = []
issue_num = 30936
keywords = []
message_count = 7.0
messages = ['298391', '298406', '298414', '298418', '298419', '298421', '298439']
nosy_count = 3.0
nosy_names = ['vstinner', 'serhiy.storchaka', 'Segev Finer']
pr_nums = ['2712', '2727', '2728']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'resource usage'
url = 'https://bugs.python.org/issue30936'
versions = ['Python 3.5', 'Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by 6ccb8c66-b6ca-4c37-b449-e35b11a6b9e4** [comment]
From builtbots:

test_json leaked [4, 4, 4] references, sum=12
test_json leaked [3, 3, 3] memory blocks, sum=9
1 test failed again:
    test_json

Probably unearthed by:
https://github.com/python/cpython/pull/2692
https://github.com/python/cpython/pull/2693

**Comment by vstinner** [comment]
On Python 3.6, the leaking test is:

test.test_json.test_speedups.TestEncode.test_bad_bool_args

**Comment by serhiy-storchaka** [comment]
New changeset 49f6449ef4b81537c19b82329caaf60596c516c2 by Serhiy Storchaka in branch 'master':
bpo-30936: Fix a reference leak in json when fail to sort keys. (bpo-2712)
https://github.com/python/cpython/commit/49f6449ef4b81537c19b82329caaf60596c516c2

**Comment by serhiy-storchaka** [comment]
New changeset a819e5e1e66a1a5f9a7073c8a1ff3c3f304c917b by Serhiy Storchaka in branch '3.6':
[3.6] bpo-30936: Fix a reference leak in json when fail to sort keys. (GH-2712). (bpo-2727)
https://github.com/python/cpython/commit/a819e5e1e66a1a5f9a7073c8a1ff3c3f304c917b

**Comment by serhiy-storchaka** [comment]
New changeset 45c471c04553d7e4cb5c328ea1ab02d6716f2bab by Serhiy Storchaka in branch '3.5':
[3.5] bpo-30936: Fix a reference leak in json when fail to sort keys. (GH-2712). (bpo-2728)
https://github.com/python/cpython/commit/45c471c04553d7e4cb5c328ea1ab02d6716f2bab

**Comment by serhiy-storchaka** [comment]
Thank you for opening this issue Segev.

Tests added in bpo-30911 exposed existing leak in 3.5 and 3.6. It was not exposed in 3.7 due to different code for sort_keys. But there was a leak when sorting keys is failed. I added a new test for this case.

2.7 is not affected because C speedup is not implemented for sort_keys in 2.7.

**Comment by vstinner** [comment]
Thanks for the fixes. I checked manually 2.7, 3.5, 3.6 and master branches with "./python -m test -R 3:3 test_json": no more leak!
